### PR TITLE
[WIP] Dashboard: DDS Injection

### PR DESCRIPTION
**Pull request by Spaqin**
# ARTIQ Pull Request

## Description of Changes

This PR adds ability to set DDS channel frequency in the dashboard. Works with AD9914, AD9912, AD9910.

It works by generating a kernel with given frequency and sending it by content to the device.

The widget for each channel displays the frequency as before. If you move your cursor over one, OVR button will pop up:

Screenshot from 2022-06-06 12-11-13

And if you press it, you will be able to change the frequency, and either apply or cancel the changes. Using Enter keys to apply and Esc to cancel works as well.

Screenshot from 2022-06-06 12-11-40

Tested with kc705 (looking into logs, kernel is run, not verified AD9914 functionality), and a Kasli with an AD9912. The core functionality works - frequency can be easily changed - but has some limitations:
- new frequency is not displayed in the menu (moninj does not report it back, I suppose?). I suppose displaying the user value doesn't make sense as it doesn't have to exactly match the reality (device can be restarted, and also see below).
- since they're all newly generated kernels, if you set one channel of an Urukul card, the rest are re-initialized and become 0 again. This would require changes in the Urukul driver to retrieve config and keep previous state somehow, instead of starting from scratch every time. Thus at this point only one channel per card can be really used.

Thus before merging I'd like to address these issues and discuss if and what should be done within this one (or maybe another one, to keep issues separate?) - that's why even though the feature itself feels complete, I'm marking it as WIP.

### Related Issue

Related #1142

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### All Pull Requests

- [x] Use correct spelling and grammar.

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

**Comment by Spaqin** [comment]
Changed the interface a bit:
Screenshot from 2022-06-07 13-27-19
Screenshot from 2022-06-07 13-27-27

(should we add an "off" button next to set? although it would be Urukul only)

Also using ``core_cache`` to write urukul CPLD config now, so multiple channels can be set and they won't switch others off.

**Comment by sbourdeauducq** [comment]
> should we add an "off" button next to set? although it would be Urukul only

Sounds like a good idea.

**Comment by sbourdeauducq** [comment]
Make sure it works in 1-EEM and 2-EEM modes.

File: artiq/dashboard/moninj.py [hunk 1]
```diff
@@ -168,15 +169,146 @@ def sort_key(self):
...
+class _DDSWidget(QtWidgets.QFrame):
+    class CancellableLineEdit(QtWidgets.QLineEdit):
+        def escapePressedConnect(self, cb):
+            self.esc_cb = cb
+
+        def keyPressEvent(self, event):
+            key = event.key()
+            if key == QtCore.Qt.Key_Escape:
+                self.esc_cb(event)
+            QtWidgets.QLineEdit.keyPressEvent(self, event)
```
Comment (sbourdeauducq) [inline artiq/dashboard/moninj.py:181]: This doesn't need to be defined inside ``_DDSWidget`` - move out?

File: artiq/dashboard/experiments.py [hunk 2]
```diff
@@ -703,6 +703,26 @@ def submit(self, expurl):
             scheduling["priority"], scheduling["due_date"],
             scheduling["flush"]))
 
+    def submit_by_content(self, content, class_name, description=None):
```
Comment (sbourdeauducq) [inline artiq/dashboard/experiments.py:706]: The experiment manager is dealing with the experiment GUI docks, the repos, etc.
It is not the place to define unrelated internal APIs.
Pass the scheduler connection to moninj instead and move that function there. ``_submit_task`` is a 2-line function that is not interesting to share.

**Pull request merged**